src/EditorFeatures/Core/Implementation/IntelliSense/AsyncCompletion/ItemManager.cs

```diff
@@ -0,0 +1,464 @@
+            var filterReason = Helpers.GetFilterReason(roslynTrigger);
+
+            var initialListOfItemsToBeIncluded = new List<ExtendedFilterResult>();
+            foreach (var item in data.InitialSortedList)
+            {
+                cancellationToken.ThrowIfCancellationRequested();
+
+                if (needToFilter && ShouldBeFilteredOutOfCompletionList(item, selectedFilters))
+                {
+                    continue;
+                }
+
+                if (!item.Properties.TryGetProperty(CompletionSource.RoslynItem, out RoslynCompletionItem roslynItem))
+                {
+                    roslynItem = RoslynCompletionItem.Create(
+                        displayText: item.DisplayText, 
+                        filterText: item.FilterText, 
+                        sortText: item.SortText,
+                        displayTextSuffix: item.Suffix);
+                }
+
+                if (Session.MatchesFilterText(_completionHelper, roslynItem, filterText, initialRoslynTriggerKind, filterReason, _recentItems))
+                {
+                    initialListOfItemsToBeIncluded.Add(new ExtendedFilterResult(item, new FilterResult(roslynItem, filterText, matchedFilterText: true)));
+                }
+                else
+                {
+                    // The item didn't match the filter text.  We'll still keep it in the list
+                    // if one of two things is true:
+                    //
+                    //  1. The user has only typed a single character.  In this case they might
+                    //     have just typed the character to get completion.  Filtering out items
+                    //     here is not desirable.
+                    //
+                    //  2. They brough up completion with ctrl-j or through deletion.  In these
+                    //     cases we just always keep all the items in the list.
+                    if (roslynTrigger.Kind == CompletionTriggerKind.Deletion ||
+                        roslynTrigger.Kind == CompletionTriggerKind.Invoke ||
+                        filterText.Length <= 1)
+                    {
+                        initialListOfItemsToBeIncluded.Add(new ExtendedFilterResult(item, new FilterResult(roslynItem, filterText, matchedFilterText: false)));
+                    }
+                }
+            }
+
+            // If the session was created/maintained out of Roslyn, e.g. in debugger; no properties are set and we should use data.Snapshot.
+            // However, we prefer using the original snapshot in some projection scenarios.
+            if (!session.Properties.TryGetProperty(CompletionSource.TriggerSnapshot, out ITextSnapshot snapshotForDocument))
+            {
+                snapshotForDocument = data.Snapshot;
+            }
+
+            var document = snapshotForDocument.GetOpenDocumentInCurrentContextWithChanges();
```

Comment:
Is document here actually being used to fetch trees? Because the WithChanges form is potentially expensive to use if this is being used from an older snapshot. It looks like you don't actually need it, so you might want a cheaper option.

Reply:
Thank you!
I hope 
var document = snapshotForDocument.TextBuffer.AsTextContainer().GetOpenDocumentInCurrentContext();
is cheaper. Please let me know if there are better options.

---
In reply to: [234839178](https://github.com/dotnet/roslyn/pull/29016#discussion_r234839178) [](ancestors = 234839178)

Reply:
Yep, that's the cheaper option.